perf_bench_test.go

```diff
@@ -24,127 +24,346 @@ package main
+			"image_rotate": "vhiveease/image_rotate:var_workload",
+			"json_serdes":  "vhiveease/json_serdes:var_workload",
+			"lr_serving":   "vhiveease/lr_serving:var_workload",
+			"cnn_serving":  "vhiveease/cnn_serving:var_workload",
+			"rnn_serving":  "vhiveease/rnn_serving:var_workload",
+		}
+		funcs  = strings.Split(*funcNames, ",")
+		result []string
+	)
+
+	for _, funcName := range funcs {
+		imageName, isPresent := images[funcName]
+		require.True(t, isPresent, "Function %s is not supported", funcName)
+		result = append(result, imageName)
 	}
 
-	log.Debugf("vmID %s: returned in %f seconds", vmIDString, time.Since(start).Seconds())
+	return result
 }
 
-func getImages() map[string]string {
-	return map[string]string{
-		"helloworld": "vhiveease/helloworld:var_workload",
-		// "chameleon":    "vhiveease/chameleon:var_workload",
-		// "pyaes":        "vhiveease/pyaes:var_workload",
-		// "image_rotate": "vhiveease/image_rotate:var_workload",
-		// "json_serdes":  "vhiveease/json_serdes:var_workload",
-		// "lr_serving":   "vhiveease/lr_serving:var_workload",
-		// "cnn_serving":  "vhiveease/cnn_serving:var_workload",
-		// "rnn_serving":  "vhiveease/rnn_serving:var_workload",
-		// "lr_training":  "vhiveease/lr_training:var_workload",
+func dumpMetrics(t *testing.T, metrics []map[string]float64, outfile string) {
+	outFile := getOutFile(outfile)
+
+	f, err := os.OpenFile(outFile, os.O_CREATE|os.O_WRONLY, 0666)
+	require.NoError(t, err, "Failed opening file")
+	defer f.Close()
+
+	headerSet := make(map[string]bool)
+	for _, metric := range metrics {
+		for area := range metric {
+			headerSet[area] = true
+		}
 	}
+
+	var headers []string
+	for area := range headerSet {
+		headers = append(headers, area)
+	}
+
+	writer := csv.NewWriter(f)
+
+	err = writer.Write(headers)
+	require.NoError(t, err, "Failed writting file")
+	writer.Flush()
+
+	for _, metric := range metrics {
+		var data []string
+		for _, header := range headers {
+			value, isPresent := metric[header]
+			if isPresent {
+				vStr := strconv.FormatFloat(value, 'f', -1, 64)
+				data = append(data, vStr)
+			} else {
+				data = append(data, "")
+			}
+		}
+		err = writer.Write(data)
+		require.NoError(t, err, "Failed writting file")
+		writer.Flush()
+	}
+}
+
+// getCPUIntenseRPS returns the number of requests per second that stress CPU for each image.
+func getCPUIntenseRPS() int {
+	var (
+		sum, result int
+		values      []int
+		funcs       = strings.Split(*funcNames, ",")
+		reqsPerSec  = map[string]int{
+			"helloworld":   10000,
+			"chameleon":    600,
+			"pyaes":        10000,
+			"image_rotate": 600,
+			"json_serdes":  600,
+			"lr_serving":   5000,
+			"cnn_serving":  20,
+			"rnn_serving":  600,
+		}
+	)
+
+	for _, funcName := range funcs {
+		values = append(values, reqsPerSec[funcName])
+		sum += reqsPerSec[funcName]
+	}
+
+	for _, rps := range values {
+		result += rps * rps / sum
+	}
+
+	return result
+}
+
+func getWarmupTime() int {
```

Comment:
this function doesn't serve its purpose. should you remove it since you added a runtime argument?

Reply:
Done